pkg/provisioning/provision.go

```diff
@@ -0,0 +1,345 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provisioning
+
+import (
+	"context"
+	"io/ioutil"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/foundation/database"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/foundation/multierror"
+	"github.com/conduitio/conduit/pkg/foundation/rollback"
+	"github.com/conduitio/conduit/pkg/pipeline"
+	"github.com/conduitio/conduit/pkg/processor"
+)
+
+type Service struct {
+	db               database.DB
+	logger           log.CtxLogger
+	pipelineService  PipelineService
+	connectorService ConnectorService
+	processorService ProcessorService
+}
+
+func NewService(
+	db database.DB,
+	logger log.CtxLogger,
+	plService PipelineService,
+	connService ConnectorService,
+	procService ProcessorService,
+) *Service {
+	return &Service{
+		db:               db,
+		logger:           logger.WithComponent("provision"),
+		pipelineService:  plService,
+		connectorService: connService,
+		processorService: procService,
+	}
+}
+
+func (s *Service) ProvisionConfigFile(ctx context.Context, path string) error {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	before, err := Parse(data)
+	if err != nil {
+		return err
+	}
+
+	got := EnrichPipelinesConfig(before)
+
+	var multi error
+	for k, v := range got {
+		err = ValidatePipelinesConfig(v)
+		if err != nil {
+			multi = multierror.Append(multi, cerrors.Errorf("invalid pipeline config: %w", err))
+			continue
+		}
+		err = s.provisionPipeline(ctx, k, v)
+		if err != nil {
+			// should we return, or add one to the multierror
+			return err
+		}
+	}
+
+	return multi
+}
+
+func (s *Service) provisionPipeline(ctx context.Context, id string, config PipelineConfig) error {
```

Comment:
It would be nice to see this function shorter/refactored, so we can quickly see what's happening at a high level.